pkg/virt-operator/creation/components/deployments.go

```diff
@@ -594,11 +643,38 @@ func NewOperatorDeployment(namespace string, repository string, imagePrefix stri
 				Value: virtLauncherShaEnv,
 			},
 		}
+
 		env := deployment.Spec.Template.Spec.Containers[0].Env
 		env = append(env, shaSums...)
 		deployment.Spec.Template.Spec.Containers[0].Env = env
 	}
 
+	if kubernetesServiceHost != "" {
```

Comment:
@mlsorensen 

practically, won't this kubernetesServiceHost variable never be set to `""` since the kubelet on most clusters ensures this env var is injected into all containers?

Reply:
I don't think so. If you do 'kubectl get deployment' you never see KUBERNETES_SERVICE_HOST. It's injected into the container definition between Kubelet and the container runtime, it isn't in the Kubernetes API unless explicitly set, right?

Reply:
My understanding and observations is that the Kubelet actually injects *all* services that the container can see as env vars <service_name>_SERvICE_HOST at the container runtime level, and these don't make it back to etcd.

I agree that if it's going to change behavior at all for current use cases, this PR shouldn't be considered. My intention is that while it's an obscure use case that requires this, it may be considered for including if it does no harm and has unit testing around it to ensure it's doing what's intended.

Reply:
Ah, I see what you're saying. The issue is the os.getEnv that virt-operator uses. I should be fetching the value from the virt-operator deployment in the API instead.

That aside, is this something that would be considered, or should I abandon this attempt to add the configuration functionality to virt-operator?

Reply:
> That aside, if it's addressed is this something that would be considered, or should I abandon this attempt to add the configuration functionality to virt-operator?

I think we can support you all here.

When i enter into a random pod on my dev cluster, I see this.

```
$ printenv  | grep KUBER
KUBERNETES_PORT_443_TCP_PROTO=tcp
KUBERNETES_PORT_443_TCP_ADDR=10.96.0.1
KUBERNETES_PORT=tcp://10.96.0.1:443
KUBERNETES_SERVICE_PORT_HTTPS=443
KUBERNETES_PORT_443_TCP_PORT=443
KUBERNETES_PORT_443_TCP=tcp://10.96.0.1:443
KUBERNETES_SERVICE_PORT=443
KUBERNETES_SERVICE_HOST=10.96.0.1
```
to me, it feels a little awkward for us to single out two of those automatically generated kubernetes environment variables and decide that we should propagate them alone.

I could be sold on the idea of a generic env var propagation mechanism though, that you all could leverage for your specific use case.

What would you think about a new environment variable called something like ```OPERATOR_ENV_VAR_PROPAGATION_WHTIELIST``` (maybe you can think of a better name).  You'd set that env var on the KubeVirt operator deployment when you want to force propagate certain variables across all components. The operator would then detect that whitelist internally and apply the corresponding key/val pairs to the pod templates for all installed components. basically, what you have now in this pr except generic.

For example, for your use case you'd set the following env vars on the virt-operator's deployment.

```
KUBERNETES_SERVICE_HOST="something"
KUBERNETES_SERVICE_PORT="someport"
OPERATOR_ENV_VAR_PROPAGATION_WHITELIST="KUBERNETES_SERVICE_HOST,KUBERNETES_SERVICE_PORT"
```

virt-operator would detect the whitelist internally and apply those key/val pairs to all the corresponding pod templates.

Reply:
Thanks for the help. That's an interesting idea, to grab and pass any of them. I was sticking to the vars that the Kubernetes client config required, I'm not sure what the others are used for but since it injects this for all Kubernetes services I suspect they're just spraying a variety of formats for the service details for developers to use.

There's still the potential issue you raise of the extra noise, sending the auto-injected vars into the Kubernetes API for all of the generated components regardless.

Take a look at the latest push. In the meantime I had worked out and tested a version that would fetch the env vars explicitly set at the API for the virt-operator deployment and send those along, rather than relying on the running environment of virt-operator. If no vars are found on the deployment spec, it doesn't create any of the other components with them. If they are on the deployment spec at the point that the install strategy is generated, they get passed down.

Reply:
> There's still the potential issue you raise of the extra noise, sending the auto-injected vars into the Kubernetes API for all of the generated components regardless. I'm happy to code it up though, if that's the suggested route that the kubevirt team wants to support.

@mlsorensen Roman and I spent quite a bit of time hashing out this issue earlier this morning. Ultimately, what we want is to support the ability to apply custom patches to any component. This would solve your use case as well as anyone else's use case where things like deployment configs need to be modified. unfortunately as we explored that approach it turns out to be quite complicated to implement. We'll do this eventually, but the design and implementation will take time.

So, long story short, i just wanted you to be aware of our thought process around the future of this sort of thing.

For now, in order to meet your use case, Roman and I think an iteration of my white list idea will work. Here's the design we landed on... Any env var passed into the operator with the `KV_IO_EXTRA_ENV` prefix will get propagated to all deployments.

For you, it would work by adding the following env vars to your virt-operator deployment.

```
KUBERNETES_SERVICE_HOST="something"
KUBERNETES_SERVICE_PORT="someport"
KV_IO_EXTRA_ENV_KUBERNETES_SERVICE_HOST="something"
KV_IO_EXTRA_ENV_KUBERNETES_SERVICE_PORT="someport"
```

The variables with the `KV_IO_EXTRA_ENV` prefix will have that prefix stripped from the key, and the corresponding key/value pairs added directly to all the deployments/daemonsets. Since your use case requires virt-operator to have these env vars as well, you'd also have to add the vars without the prefix for virt-operator to consume correctly which is why there's a bit of redundancy in the example. 

any thoughts on this approach?

Reply:
That will work, and I'm not in any position to quibble about any small implementation details! I appreciate the consideration. I should be able to adjust and reuse my existing implementation without too much trouble.

I do like the idea of being able to patch arbitrary things. Potential things people may need to adjust are things like resource limits, maybe additional labeling, etc. It would be easy to fix any random sticky situations. Should I open an issue for that, or is there something already on the list?

Reply:
> I should be able to adjust and reuse my existing implementation without too much trouble.

great!

>  Should I open an issue for that, or is there something already on the list?

i think we need an issue. i haven't seen this idea of a generic approach to patching virt-operator related resources come up officially yet until now.

Reply:
Hi @davidvossel, I guess I'll follow up on this particular thread. I abandoned the previous iteration that grabs env vars from the operator's definition and implemented what you described, then rebased on master and force pushed it here to keep things clean.

One small variance from the specification you described - the virt-operator will detect "KV_IO_EXTRA_ENV_", strip the prefix, and turn around and set those vars in its own environment as well prior to creating its clientset, so we don't have to duplicate the KUBERNETES_SERVICE_HOST in the operator. This way all of the KV_IO_EXTRA_ENV will be applied equally as their non "KV_IO_EXTRA_ENV_" variables across virt-operator, virt-api, virt-controller, and virt-handler.

Reply:
works for me